stdlib/public/Concurrency/ExecutorAssertions.swift

```diff
@@ -11,6 +11,134 @@
 //===----------------------------------------------------------------------===//
 
 import Swift
+import SwiftShims
+
+// ==== -----------------------------------------------------------------------
+// MARK: Precondition executors
+
+/// Unconditionally if the current task is executing on the expected serial executor,
+/// and if not crash the program offering information about the executor mismatch.
+///
+/// This function's effect varies depending on the build flag used:
+///
+/// * In playgrounds and `-Onone` builds (the default for Xcode's Debug
+///   configuration), stops program execution in a debuggable state after
+///   printing `message`.
+///
+/// * In `-O` builds (the default for Xcode's Release configuration), stops
+///   program execution.
+///
+/// * In `-Ounchecked` builds, the optimizer may assume that this function is
+///   never called. Failure to satisfy that assumption is a serious
+///   programming error.
+///
+/// - Parameter executor: the expected current executor
+@available(SwiftStdlib 5.9, *) // FIXME: use @backDeploy(before: SwiftStdlib 5.9)
+public
+func preconditionTaskOnExecutor(
+    _ executor: some SerialExecutor,
+    message: @autoclosure () -> String = String(),
+    file: StaticString = #fileID, line: UInt = #line
+) {
+  let expectationCheck = _taskIsCurrentExecutor(executor.asUnownedSerialExecutor().executor)
+
+  /// TODO: implement the logic in-place perhaps rather than delegating to precondition()?
+  precondition(expectationCheck,
+      // TODO: offer information which executor we actually got
+      "Incorrect actor executor assumption; Expected '\(executor)' executor. \(message())",
+      file: file, line: line) // short-cut so we get the exact same failure reporting semantics
```

Comment:
The implementation bodies are suboptimal here, we'll do better including offering the "here's the executor we found instead", but for now at least we can have the functionality and APIs and then we can improve their internal sand messages.